Add `hpc-repo-tools` to repository

Onitoxan opened the pull request:
> [!Warning]
> When rebasing remember to update `.git-blame-ignore-revs` commits SHA for mass formatting and ESlint automatic fixes

## Nature of PR
- Bug-fix
- Hot-fix
- Feature
- Testing
- Rewrite
- CI
- 🟢  Dependencies

## Description

In this PR we focus on adding latest version of `hpc-repo-tools` to `hpc-cdm`. This means we have to update code to meet ESlint rules we specify and format all files again. Beside this we also migrate to flat config in ESlint.

## How to test changes

For testing, you can run this command for each project if you want to test individually
```
npx nx run {projectName}:lint

// If you want to run them all:
npm run lint
```

Beside this, ESlint changes shouldn't have broken the app, what its unlikely, but is good to check.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on package.json:
```diff
@@ -70,8 +70,9 @@
-    "@typescript-eslint/parser": "7.18.0",
+    "@typescript-eslint/eslint-plugin": "8.31.0",
+    "@typescript-eslint/parser": "8.31.0",
+    "@unocha/hpc-repo-tools": "6.4.0",
```
There is newer version of repo-tools now, let's install the newest version.

Pl217 commented on package.json:
```diff
@@ -70,8 +70,9 @@
     "@types/styled-components": "5.1.26",
-    "@typescript-eslint/eslint-plugin": "7.18.0",
-    "@typescript-eslint/parser": "7.18.0",
+    "@typescript-eslint/eslint-plugin": "8.31.0",
+    "@typescript-eslint/parser": "8.31.0",
```
Remove `@typescript-eslint/*` packages, because they are now combined into `typescript-eslint`, which comes with repo-tools

Pl217 commented on package.json:
```diff
@@ -82,12 +83,12 @@
     "enketo-core": "^5.16.6",
-    "eslint": "8.57.0",
+    "eslint": "9.23.0",
     "eslint-config-prettier": "10.1.1",
```
`eslint-config-prettier` comes with repo-tools

Pl217 commented on package.json:
```diff
@@ -82,12 +83,12 @@
     "eslint-plugin-import": "2.31.0",
     "eslint-plugin-jsx-a11y": "6.10.1",
-    "eslint-plugin-react": "7.32.2",
+    "eslint-plugin-react": "7.37.5",
```
I wonder whether we need these plugins at all, since they are not referenced directly anywhere. Remove all `eslint-plugin/*` dependencies if they are not needed.

Also, I will use this comment to ask whether both `@nx/eslint` and `@nx/eslint-plugin` are needed? Maybe the `eslint-plugin/*` dependencies are implicitly used by this `@nx/eslint`, but you should investigate how they piece together and clean up.

Onitoxan replied:
So indeed they are needed for `@nx/eslint-plugin` this `eslint-plugin-*` dependencies. They are not listed as peer dependencies (What I don't understand why wouldn't they do it). But if you go to `node_modules/@nx/eslint-plugin/src/flat-configs/react-base.js` as example, they require:

```js
const importPlugin = tslib_1.__importStar(require("eslint-plugin-import"));
```

And to answer to your second question, I feel for the moment is nice to have `nx/eslit-plugin` since there are specific react rules that are helpful, like rules for hooks, JSX eslint rules... If you want, we could start defining React ESlint rules in `hpc-repo-tools` so we are always in control of what is being applied. But for the moment I would keep them until the day where we define rules for React in `hpc-repo-tools` arrives.

Honestly, it would be helpful to have it in `hpc-repo-tools`. `@nx/eslint-plugin` install rules also for Angular and other stuff that we don't really need.

Long story short: Lets keep all this ESlint dependencies until we can get rid of `@nx/eslint-plugin`

Pl217 replied:
You are right about these plugins being used by `@nx/eslint-plugin`. You did specify the path for `eslint-plugin-import`, and I'd add that `eslint-plugin-jsx-a11y`, `eslint-plugin-react` and `eslint-plugin-react-hooks` are used in `node_modules/@nx/eslint-plugin/src/flat-configs/react-jsx.js`. That means `eslint-plugin-cypress` is not used, but I will talk about that separately.

Regarding moving React specific rules to `hpc-repo-tools`, I'd be against that. I want that repo to host rules applicable to every other repo where we use it. Each repo can then add rules on top (that's why we call it `baseConfig`). And React rules are pretty specific and not generally applicable. And I especially don't want to install `@nx/eslint-plugin` in repo-tools, since React-specific rules we're adding are actually coming from that package.

Pl217 commented on jest.config.ts:
```diff
@@ -1,3 +1,3 @@
+import { getJestProjectsAsync } from '@nx/jest';
 
-export default { projects: getJestProjects() };
+export default { projects: await getJestProjectsAsync() };
```
Explain this change. You already committed it separately, which is good, but I want to know more about why we are moving to async version.

Onitoxan replied:
ESlint rule was asking to go for this change, so I just followed it 🤔

Onitoxan replied:
Just checked again, its a deprecated method `getJestProjectsAsync` it will be removed in `@nx v21` and they ask to move to this async method

Pl217 commented on tsconfig.base.json:
```diff
@@ -8,10 +8,10 @@
     "experimentalDecorators": true,
     "importHelpers": true,
-    "target": "es2015",
+    "target": "es2021",
```
Change of `target` is pretty significant and should at least be committed separately, but if not needed here (which I think it isn't), then do in separate PR.

Onitoxan replied:
If I recall correctly, what motivated this change of `lib` (so I also decided to up `target`) it was a ESlint rule that asked to go from `String.prototype.replace()` to `String.prototype.replaceAll()` and this one it's only available from `ES2021`.

I should commit it in a separate commit, but what do you think of beside this update, to upgrade to `ES2024` in another PR? We could also add `ESNext.Iterator` and `ESNext.Collections`

Pl217 replied:
I'd like to increase the `lib`, but increase of `target` depends on browser support, which is a question for product owners.

Pl217 reviewed (changes requested):
1. There are still references to `eslintrc.json` in `nx.json`
2. Don't combine config changes and fixes to files, because it will be easier for me to look at them separate. It also makes sense logically to separate them. Combining such changes in same commit may only benefit if we want to have each commit able to compile, when doing bisect, but in this case with linting fixes, they don't impede compilation, so there is no reason to have everything in a single commit
3. Split linting fixes into auto-fixes and those manually fixed. This will make it easier for me to review (I can autopilot auto-fixed errors and pay more attention to manually fixed ones), but also allow to ignore revision with auto-fixes
4. Add linting to `lint-staged`, not only formatting. In formatting configuration, you may freely pick all files with `*`. Check if something needs to be added to `.prettierignore`
5. Check entire project for formatting in CI
6. Add `.git-blame-ignore-revs` and configure it in `prepare` script. Populate the file with revisions where you do mass formatting and auto-fixing of linting errors
7. Rework this and commit differently. I think it's easier to start from scratch and do things one-by-one and commit them into those logical areas I outlined above. Easier compared to trying to amend the current commits

Onitoxan committed:
Add `hpc-repo-tools`

Onitoxan committed:
Update to new prettier and eslint flat config

Onitoxan committed:
Run `eslint --fix`

Onitoxan committed:
Fix ESlint errors manually

Onitoxan committed:
Run `prettier -w`

Onitoxan committed:
Replace import `FormObjectValue` for `util`

On previous commit, prettier removed the export
of `FormObjectValue` for being a duplication of
`export * as util`. Therefore we update her for
this change

Onitoxan committed:
Increase `lib` to `ES2021`

This change is motivated by the need of
`String.prototype.replaceAll()`

Onitoxan committed:
Setup `.git-blame-ignore-revs`

Onitoxan committed:
Configure `lint-staged` to check all files

Check all files except the ones written in
.prettierignore

Onitoxan committed:
Replace deprecated `getJestProjects()`

`getJestProjects()` is removed in `@nx` `v21`

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on package.json:
```diff
@@ -82,14 +81,13 @@
     "enketo-core": "^5.16.6",
-    "eslint": "8.57.0",
-    "eslint-config-prettier": "10.1.1",
+    "exceljs": "4.4.0",
```
Why did you move `exceljs` up here? It should be the last package starting with letter E. Look at `package-lock.json`, because it sorts everything in default order

Pl217 commented on eslint.config.mjs:
```diff
@@ -0,0 +1,45 @@
+
+export default [
+  ...baseConfig,
+  { plugins: { '@nx': nxEslintPlugin } },
```
I actually didn't know you could specify plugins in flat config like this. I have set up a new project with latest NX v21 and in top-level `eslint.config.mjs` file that it produced, `@nx/eslint-plugin` is used like this:
```js
import nx from '@nx/eslint-plugin';

export default [
  ...nx.configs['flat/base'],
  ...nx.configs['flat/typescript'],
  ...nx.configs['flat/javascript'],
  // ...
]
```
This would actually be my preferred way because I know what's happening, you just spread another config. With `plugins` declaration, I don't know how it works. Look into these different ways of using the plugin and see whether we really need all three flat configs being spreaded.

Pl217 commented on eslint.config.mjs:
```diff
@@ -0,0 +1,45 @@
+        'error',
+        {
+          enforceBuildableLibDependency: true,
+          allow: [],
```
In the fresh setup of Nx project, this is how this line looks:
```suggestion
          allow: ['^.*/eslint(\\.base)?\\.config\\.[cm]?js$'],
```

Pl217 commented on eslint.config.mjs:
```diff
@@ -0,0 +1,45 @@
+      'jest.preset.js',
+      'prettier.config.mjs',
+      'tailwind.config.js',
+      'dist',
```
In the fresh setup of Nx project, this is how this line looks:
```suggestion
      '**/dist',
```

Pl217 commented on apps/hpc-cdm-e2e/eslint.config.mjs:
```diff
@@ -0,0 +1,15 @@
+  ...baseConfig,
+  {
+    ignores: ['**/dist'],
+  },
```
I see you do this here in CDM E2E, but didn't do in CDM app. If you add this in top-level config (as I wrote in other comment), there will be no need to specify everywhere else. I don't know if there is `dist` anywhere else besides top-level, so this may be useless, but still, do it once in top-level config.

Pl217 commented on apps/hpc-cdm-e2e/eslint.config.mjs:
```diff
@@ -0,0 +1,15 @@
+      '@typescript-eslint/no-var-requires': 'off',
+      'no-undef': 'off',
+    },
+  },
```
This file was deleted in d49762f2c1b2ba549965063610b7138d24f3a6dd

Pl217 commented on apps/hpc-cdm-e2e/eslint.config.mjs:
```diff
@@ -0,0 +1,15 @@
+import baseConfig from '../../eslint.config.mjs';
+
+export default [
+  ...baseConfig,
```
You didn't add Cypress plugin here, so it is not used anywhere in the project. If you look at deleted `.eslintrc.json`, you can see it was used before. This is how E2E config looks in fresh Nx project:

```suggestion
import nx from '@nx/eslint-plugin';
import baseConfig from '../../eslint.config.mjs';

export default [
  ...baseConfig,
  ...nx.configs['flat/react'],
```

Pl217 commented on apps/hpc-ftsadmin-e2e/eslint.config.mjs:
```diff
@@ -0,0 +1,7 @@
+  {
+    ignores: ['**/dist'],
+  },
+];
```
Applying the same I said for CDM E2E, this file should be:

```suggestion
import nx from '@nx/eslint-plugin';
import baseConfig from '../../eslint.config.mjs';

export default [
  ...baseConfig,
  ...nx.configs['flat/react'],
];
```

Pl217 commented on libs/hpc-ui/eslint.config.mjs:
```diff
@@ -0,0 +1,16 @@
+  },
+  {
+    ignores: ['**/postcss.config.js', '**/jest.config.ts'],
+  },
```
Out of all lib configs, this is the only one where you're ignoring `jest.config.ts`. I'd say you should do it globally in top-level config.

Pl217 commented on libs/hpc-live/.eslintrc.json:
You didn't add flat config in `hpc-live` lib. This may be on purpose as you've been adding them in other libs to turn off rules.

Onitoxan replied:
Yeah it was intentional since there was no extra rule then the base config

Pl217 commented on apps/hpc-cdm/src/app/components/enketo/overrides/dialog.ts:
```diff
@@ -11,6 +11,7 @@ type DialogContentObject = {
 
 // TODO: implement prompts at some point
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
 const alert = (content: string | DialogContentObject) => Promise.reject('TODO');
```
These are resolved by putting an underscore in front of a variable. But, in this case, I'd just remove it and `prompt` below, as they aren't used anywhere and this TODO will likely never be TODONE
```suggestion
const alert = (_content: string | DialogContentObject) => Promise.reject('TODO');
```

Pl217 commented on apps/hpc-cdm/src/app/components/enketo/xform.ts:
```diff
@@ -31,22 +31,22 @@ export default class XForm {
+    const { onDataUpdate, onPageFlip } = opts ?? {};
     this.files = files;
 
+    // eslint-disable-next-line require-await
```
Instead of disabling the rule, you can remove `async` from the method and wrap each return with `Promise.resolve()`. The first return should be `Promise.resolve(undefined)` for example.

Pl217 commented on apps/hpc-cdm/src/app/utils/usePrompt.tsx:
```diff
@@ -30,18 +30,20 @@ function useConfirmExit(confirmExit: () => boolean, when = true) {
 export function usePrompt(message: string, when = true) {
   useEffect(() => {
     if (when) {
+      // eslint-disable-next-line unicorn/prefer-add-event-listener
```
What is the reason this linting rule is suppressed?

Pl217 commented on apps/hpc-cdm/src/polyfills.ts:
```diff
@@ -3,5 +3,6 @@
  *
  * See: https://github.com/zloirock/core-js#babel
  */
+// eslint-disable-next-line unicorn/no-unnecessary-polyfills
```
I think we should configure this rule instead of suppressing the warning. Linter picks up `engines` from `package.json`, which specify Node.js 22+, but we actually target browsers, not Node, so we should configure this rule with browser compatibility in mind.

Onitoxan replied:
I configured the rule like so:

```js
 rules: {
      /**
       * Same as in `.browserslistrc`
       */
      'unicorn/no-unnecessary-polyfills': [
        'error',
        {
          targets: '> 0.2% and not dead and not op_mini all',
        },
      ],
    },
```

Is this correct? If so, I still get the ESlint error to remove this polyfill, I tried more weird browserlist query like `0.1% and ie 11` and if I use it it doens't complain about the polyfill.

Since you may know more, I ask you, should I then remove the polyfill or is it badly configured  `'unicorn/no-unnecessary-polyfills'` ?

Pl217 replied:
I looked at [rule docs](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/3646efbd14979ca3dba438325aec2ffac33fa081/docs/rules/no-unnecessary-polyfills.md) and tried:
`'unicorn/no-unnecessary-polyfills': ['error', { targets: 'chrome 1' }]`
vs
`'unicorn/no-unnecessary-polyfills': ['error', { targets: 'chrome 100' }]`

Then I removed your comment from `apps/hpc-ftsadmin/src/polyfills.ts` that disables the rule and tested both with: `npx eslint --quiet apps/hpc-ftsadmin/src/polyfills.ts`.

Turns out, with Chrome 100, we don't need the polyfills, with Chrome 1, we do need them. So, getting linting errors is correct and you should not just suppress them out of being afraid to break things.

You can see what browsers the rules compile by using `npx browserslist --config=apps/hpc-cdm/.browserslistrc`. We don't need the polyfill for the rules in browserslist indeed.

What's more, I checked the freshly created Nx project (with `npx create-nx-workspace`) and there are no polyfills files at all, neither does it use browsers list. The `target` and `lib` in `tsconfig.base.json` are both ES2022.

I checked what happened with other polyfill (`regenerator-runtime/runtime`), and it was [being installed](https://github.com/UN-OCHA/hpc-cdm/blob/ef1fdb00b552045e8333ba7bfce6230f61d6e057/package-lock.json#L21366-L21371) as a [dependency](https://github.com/UN-OCHA/hpc-cdm/blob/ef1fdb00b552045e8333ba7bfce6230f61d6e057/package-lock.json#L2416-L2428) of `@babel/runtime`, which [removed it in `v7.27.1`](https://github.com/babel/babel/releases/tag/v7.27.1) in a commit saying:
> Remove unused `regenerator-runtime` dep in `@babel/runtime`

You decide whether you want to remove polyfills in this PR, because linter forces you or in a separate PR. But, we still have a question of browser support open, since project owners need to decide what we should support.

In a separate PR, you should remove browserslist if it isn't used by anything, which you should determine. Keep in mind that we have [this custom code](https://github.com/UN-OCHA/hpc-cdm/blob/ef1fdb00b552045e8333ba7bfce6230f61d6e057/apps/hpc-cdm/src/app/components/enketo/util.ts#L3-L12) to check for browser versions and this list likely need updating as well.

Onitoxan replied:
I will remove `polyfills.ts` in this PR, but we should create a separate PR, as you mention, to remove `browserslist` and review the enketo browser support.

Pl217 commented on apps/hpc-ftsadmin/src/app/utils/table-headers.ts:
```diff
@@ -1,7 +1,8 @@
-import { LanguageKey, t } from '../../i18n';
-import { Strings } from '../../i18n/iface';
+/* eslint-disable unicorn/prefer-set-has */
```
I guess it was the types that you didn't know how to resolve when wrapping the below arrays in `new Set()`, you should have just used `<string>` generic with `Set` constructor

Onitoxan replied:
Mmm I mean yes and no, I was thinking because in most up to date code this bit will be removed because it was not written correctly. https://github.com/UN-OCHA/hpc-cdm/blob/HPC-8546/apps/hpc-ftsadmin/src/app/utils/table-headers.ts

If you check now everything is written with generics. What in the hypothetical case of still having to use `Set` it would not have typing errors. So I muted the rule simply because I knew this was going out. Nonetheless, I will just do `Set<string>` since anyways this will be deleted

Pl217 replied:
This is not done

Onitoxan replied:
@Pl217 Must have eaten it in a rebase, its now done.

Pl217 commented on libs/hpc-dummy/tsconfig.lib.json:
```diff
@@ -8,5 +8,5 @@
   },
   "exclude": ["**/*.spec.ts", "**/*.test.ts", "jest.config.ts"],
-  "include": ["**/*.ts"]
+  "include": ["**/*.ts", "**/*.tsx"]
```
I don't think this change belongs in PR titled "Fix ESlint errors manually"

Pl217 commented on libs/hpc-dummy/src/lib/users.tsx:
```diff
@@ -1,16 +1,16 @@
 import { ThemeProvider } from '@unocha/hpc-ui';
 
-import { User } from './data-types';
+import { type User } from './data-types';
```
Why is this change not done by auto-fixing?

Pl217 commented on libs/hpc-ui/src/lib/components/dialogs.tsx:
```diff
@@ -211,6 +209,7 @@ export const Dialogs = () => {
             </Dialog>
           );
         }
+        return null;
```
You return `null` because of linting error requiring to return a value in `Array.prototype.map()`, but I wonder what that would mean. You are returning `<Dialog> ... </Dialog>` otherwise. How would react treat when it encounters `null` in an array of components, I don't know.

If `null` is a problem, you could throw an error, since we should not be reaching this code anyway, since `d.type` is typed as `"alert" | "confirm"`.

Onitoxan replied:
With React, if you are returning an array of components, if there is `null` it means to render nothing. If you just returned `undefined` or just wrote `return` (What is the same, because returns `undefined`) this would result in the string `undefined` being rendered. So there is no issue with letting `null`

Pl217 commented on tools/remove-unneeded-deps.js:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @typescript-eslint/no-require-imports */
```
You could convert this file to `.mjs` and use imports, not `require`

Pl217 commented on apps/hpc-cdm/src/app/components/enketo/formToolbar/indicator.tsx:
```diff
@@ -120,8 +120,7 @@ const Indicator = (props: Props) => {
         </StatusDetails>
-      }
-    >
+      }>
```
If you want `>` to go on next line, like it used to, you could use Prettier config `bracketSameLine: false`.

Pl217 commented on tsconfig.base.json:
```diff
@@ -8,10 +8,10 @@
     "experimentalDecorators": true,
     "importHelpers": true,
-    "target": "es2015",
+    "target": "ES2015",
```
Commit message says "Increase `lib` and `target` to `ES2021`", but you haven't increased `target`, only lib.

Pl217 commented on .prettierignore:
```diff
@@ -1,8 +1,12 @@
+.git-blame-ignore-revs
+.editorconfig
+.npmrc
+Dockerfile
```
We have the capacity to format `Dockerfile`, why did you add it to ignore list?

Pl217 commented on .git-blame-ignore-revs:
```diff
@@ -0,0 +1,2 @@
+fcd14a0f08ebf0af88ebca2f78ea922df3d1b2cf
+59c5fe37244cea299f205e662157af798a945d0d
```
You officially hate newlines at the end of files :P

Pl217 commented on libs/hpc-core/eslint.config.mjs:
```diff
@@ -0,0 +1,11 @@
+  {
+    files: ['**/*.ts'],
+    rules: {
+      'unicorn/prefer-export-from': 'off',
```
I commented out this rule and `npx nx run hpc-core:lint` still passes

Pl217 commented on libs/hpc-data/eslint.config.mjs:
```diff
@@ -0,0 +1,12 @@
+    files: ['**/*.ts'],
+    rules: {
+      '@typescript-eslint/no-unused-vars': 'off',
+      'unicorn/prefer-export-from': 'off',
```
I commented out this rule and `npx nx run hpc-data:lint` still passes.

Pl217 commented on libs/hpc-data/eslint.config.mjs:
```diff
@@ -0,0 +1,12 @@
+  {
+    files: ['**/*.ts'],
+    rules: {
+      '@typescript-eslint/no-unused-vars': 'off',
```
You could resolve this problem by exporting variables, or prefixing them with underscore.

Pl217 commented on libs/hpc-dummy/eslint.config.mjs:
```diff
@@ -0,0 +1,11 @@
+  {
+    files: ['**/dummy.ts'],
+    rules: {
+      'require-await': 'off',
```
You could have also disabled this in file `dummy.ts`

Pl217 commented on libs/hpc-ui/eslint.config.mjs:
```diff
@@ -0,0 +1,16 @@
+  ...baseConfig,
+  {
+    rules: {
+      'unicorn/prefer-export-from': 'off',
```
I have commented out this rule and `npx nx run hpc-ui:lint` still passes.

Pl217 reviewed (changes requested):
1. Write a reminder in the first comment that we should update revisions in `.git-blame-ignore-revs` if we do a rebase
2. Set up formatting checks in CI. It will show you that you haven't formatted all files. I already mentioned that under point 5 in previous review
3. There are linting errors still, which CI shows
4. Script `lint` in `package.json` should fix formatting everywhere, not only specific file types. Use `.` instead at the end of the command

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan commented:
> 2\. Set up formatting checks in CI. It will show you that you haven't formatted all files. I already mentioned that under point 5 in previous review

I don't really get what you mean. On `workflow.yml` in step `Lint affected apps` we run `npm run lint` that its equal to: `nx affected:lint && nx lint && prettier --check '**/*.{ts,tsx,js}'`. I will make `prettier` check all files. Are you referring to move it in a separate step in CI or what do you want me to do?

Onitoxan committed:
Check formatting of all files on lint command

Onitoxan committed:
Add '.tsx' files in tsconfig for 'hpc-dummy'

Pl217 commented:
> > 2. Set up formatting checks in CI. It will show you that you haven't formatted all files. I already mentioned that under point 5 in previous review
> 
> I don't really get what you mean. On `workflow.yml` in step `Lint affected apps` we run `npm run lint` that its equal to: `nx affected:lint && nx lint && prettier --check '**/*.{ts,tsx,js}'`. I will make `prettier` check all files. Are you referring to move it in a separate step in CI or what do you want me to do?

Aha right, just do the point 4 from my previous comment, to check entire directory, not just the files with the specific extension. I saw the check `npx prettier -c .` failed when I tried it and I didn't see it in CI, but it was likely bundled with linting failures from ESLint, so I didn't pay attention. In other repos, we have separate checks for formatting and linting in CI.

Onitoxan committed:
Remove polyfills

ESlint was indicating when using `.browserlistrc`
configuration, that there was no need for
`core-js/stable`. Also, `regenerator-runtime/runtime`
was removed in v7.27.1 from babel, so there is no
need for it anymore.

https://github.com/babel/babel/releases/tag/v7.27.1

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on tools/remove-unneeded-deps.mjs:
```diff
@@ -1,14 +1,12 @@
-const path = require('path');
-
-const PACKAGE = path.join(path.dirname(__dirname), 'package.json');
+const PACKAGE = path.join(import.meta.dirname, 'package.json');
```
Have you tried running `npm run remove-unneeded-deps` after refactoring this file? It fails to find `package.json`, because `import.meta.dirname` evaluates to `tools/` directory where this script lives.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on package.json:
```diff
@@ -82,13 +81,12 @@
-    "eslint": "8.57.0",
-    "eslint-config-prettier": "10.1.1",
+    "eslint": "9.27.0",
     "eslint-plugin-cypress": "2.13.2",
```
Why didn't you upgrade `eslint-plugin-cypress` but upgraded all other ESLint plugins?

Pl217 commented on apps/hpc-cdm-e2e/eslint.config.mjs:
```diff
@@ -0,0 +1,4 @@
+import nxESlintPlugin from '@nx/eslint-plugin';
+import baseConfig from '../../eslint.config.mjs';
+
+export default [...baseConfig, ...nxESlintPlugin.configs['flat/react']];
```
I copied from wrong file when suggesting changes in [this comment](https://github.com/UN-OCHA/hpc-cdm/pull/505#discussion_r2103106941). I correctly mentioned we should use Cypress plugin though.
```suggestion
import cypress from 'eslint-plugin-cypress/flat';
import baseConfig from '../../eslint.config.mjs';

export default [cypress.configs['recommended'], ...baseConfig];
```
Do the same in other E2E project.

Pl217 reviewed (changes requested):
Why Dockerfile is not format checked? 🤔

Onitoxan commented:
> Why Dockerfile is not format checked? 🤔

I don't really get why with `.` it doesn't pick it up, it must be that `.` doesn't pick files without extension. So I decided to change it to `'**/*' '*'` so we pick up every kind of file.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on .prettierignore:
```diff
@@ -1,8 +1,19 @@
 
 /dist
 /coverage
+/env
```
We only have a single file here, which is `.conf` file. In similar situation in other repos, I excluded that file type with ignoring `*.conf` rather than ignoring the entire directory.

Pl217 commented on .prettierignore:
```diff
@@ -1,8 +1,19 @@
 /dist
 /coverage
+/env
+/.husky
```
This counters the setting you have in `prettier.config.mjs` where we indicate that files under `.husky/*` pattern should be formatted with shellscript parser. Instead, add this:
```suggestion
/.husky/_/
```

Pl217 commented on .prettierignore:
```diff
@@ -1,8 +1,19 @@
 /coverage
+/env
+/.husky
+/.github
```
Why do you want to exclude `.github` directory? We have JSON and YAML files in there, which should be formatted and checked. I guess this is because you hate newlines at end of files and wanted to prevent formatter from adding one to `.github/CODEOWNERS` :P

Pl217 commented on .prettierignore:
```diff
@@ -1,8 +1,19 @@
+**/.gitkeep
+**/.browserslistrc
+**/favicon.ico
+**/assets
```
There are some `.tsx` and `.svg` files under `/assets` directories, which we could format. I'd just exclude file types we don't support, and I've seen we need to include `.png` and `.woff` now. Also, when `apps/hpc-cdm/src/assets/logos/kobologo.svg` gets formatted, you could use "Convert indentation to spaces" in VS Code, because the file now mixes spaces and tabs.

Pl217 reviewed (changes requested):
Few comments about what we exclude from formatting checks

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on eslint.config.mjs:
```diff
@@ -0,0 +1,83 @@
+      '.nx',
+      '**/eslint.config.mjs',
+      '**/jest.config.ts',
+      '**/jest.config.ts',
```
You have the same line twice

Pl217 commented on eslint.config.mjs:
```diff
@@ -0,0 +1,83 @@
+      'prettier.config.mjs',
+      'tailwind.config.js',
+      '**/dist',
+      'nx.json',
```
I don't think JSON files can be linted

Pl217 commented on apps/hpc-ftsadmin/src/app/components/organization-form.tsx:
```diff
@@ -132,10 +135,10 @@ export const OrganizationForm = ({ initialValues, id, load }: Props) => {
-          if (errors.isDuplicateError(err)) {
-            setErrorValue(err.value);
-            setError(err.code);
+        .catch((_error) => {
```
This is not unused. Why start variable name with underscore?

Pl217 commented on apps/hpc-ftsadmin/src/app/components/filters/filter-organization-table.tsx:
```diff
@@ -1,20 +1,20 @@
+import { Form, Formik, type FormikState } from 'formik';
 import tw from 'twin.macro';
 
+import { util as codecs, type util } from '@unocha/hpc-data';
```
Why not use `codecs.FormObjectValue` like in other places? Or maybe don't put alias on `util` and just use `util` for everything everywhere.

Pl217 reviewed (changes requested):
Will we ever merge this? 😄

Would have made my life easier if you didn't commit so many files in a single commit with manual linting fixes, but if you divided it into multiple commits by type of change being done. Not asking you to do it now, just sayin.

Commit message in 75fba69acc992cd384edfc98b991f4138f1a2af6 could be better and explain the change being made.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on apps/hpc-ftsadmin/src/app/components/organization-form.tsx:
```diff
@@ -146,12 +149,12 @@ export const OrganizationForm = ({ initialValues, id, load }: Props) => {
-          if (errors.isDuplicateError(err)) {
-            setErrorValue(err.value);
-            setError(err.code);
+        .catch((_error) => {
```
Missed a spot

Pl217 commented on apps/hpc-ftsadmin/src/app/components/organization-form.tsx:
```diff
@@ -1,24 +1,27 @@
+  errors,
+  type organizations,
+  type util,
+} from '@unocha/hpc-data';
```
You have `util` twice here as well

Pl217 commented on apps/hpc-ftsadmin/src/app/components/filters/filter-flows-table.tsx:
```diff
@@ -1,18 +1,14 @@
 
+import { Alert } from '@mui/material';
+import { util } from '@unocha/hpc-core';
+import { util as codecs } from '@unocha/hpc-data';
```
If you resolve the previous comment by only keeping `util` and dropping `codecs` alias, this would be the only remaining usage, so remove it from here too.

Pl217 reviewed (changes requested):
You didn't go all the way with those changes from last review.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 reviewed (approved):
Finally
